docs/BestPractices.md

```diff
@@ -0,0 +1,56 @@
+
+### Close input stream from client operations
+
+For streaming operations such as `S3Client#getObject`,  if you are working with `ResponseInputStream` directly, we have the following recommendations:
+
+- Use the data from the input stream as soon as possible
+- Read all data from the stream
+- Close the input stream as soon as possible
+
+This is because the input stream is a direct stream of data from the HTTP connection and the underlying HTTP connection cannot be
+reused until all data from the stream has been read and the stream is closed. If these rules are not followed, the client can 
+run out of resources by allocating too many open, but unused, HTTP connections.
+
+### Tune HTTP configurations based on performance tests
+
+SDK provides a set of [default http configurations] that apply to general use cases. Customers are recommended to tune the configurations for their applications based on their use cases.
+
+### Use OpenSSL for Netty async client
+
+By default, `NettyNioAsyncHttpClient` uses JDK as the SslProvider. In our local tests, we found that using `OpenSSL`
+is more performant than JDK. Using OpenSSL is also the recommended approach by Netty community, see [Netty TLS with OpenSSL].
+Note that you need to add `netty-tcnative` to your dependency in order to use OpenSSL, see [Configuring netty-tcnative]
+
+Once you have `netty-tcnative` configured correctly, `NettyAsyncHttpClient` will select OpenSSL automatically, or you can set the
+`SslProvider` explicitly on the NettyNioAsyncHttpClient builder.
+
+```
+   NettyNioAsyncHttpClient.builder()
+                          .sslProvider(SslProvider.OPENSSL)
+                          .build();
+```
+
+### Utilize timeout configurations
+
+SDK provides timeout configurations on requests out of box and they can be configured via `ClientOverrideConfiguration#apiCallAttemptTimeout` and `ClientOverrideConfiguration#ApiCallTimeout`. See [example]
+
+- `ApiCallAttemptTimeout` tracks the amount of time for a single http attempt and the request can be retried if timed out on api call attempt. 
+- `ApiCallTimeout` configures the amount of time for the entire execution including all retry attempts. 
+
+By default, timeouts are disabled. Using them together is helpful to set a hard limit on total time spent on all attempts across retries and each individual HTTP request to fail fast on one slow request.
+
+[default http configurations]: https://github.com/aws/aws-sdk-java-v2/blob/master/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpConfigurationOption.java
+[Netty TLS with OpenSSL]: https://netty.io/wiki/requirements-for-4.x.html#tls-with-openssl
+[Configuring netty-tcnative]: https://netty.io/wiki/forked-tomcat-native.html
+[example]: https://github.com/aws/aws-sdk-java-v2/blob/master/test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/timeout/sync/SyncApiCallTimeoutTest.java#L62
```

Comment:
Links to specific lines in master branch tend to get out-of-date, maybe link to a tag instead?
Or maybe put the example in the body of the text.

Reply:
Yeah, good point. Will update